cli/azd/pkg/tools/azcli/container_registry.go

```diff
@@ -85,6 +87,27 @@ func (crs *containerRegistryService) GetContainerRegistries(
 }
 
 func (crs *containerRegistryService) Login(ctx context.Context, subscriptionId string, loginServer string) error {
+	dockerCreds, err := crs.Credentials(ctx, subscriptionId, loginServer)
+	if err != nil {
+		return err
+	}
+
+	err = crs.docker.Login(ctx, dockerCreds.LoginServer, dockerCreds.Username, dockerCreds.Password)
+	if err != nil {
+		return fmt.Errorf(
+			"failed logging into docker registry %s: %w",
+			loginServer,
+			err)
+	}
+
+	return nil
+}
+
+// Credentials gets the credentials that could be used to login to the specified container registry. It prefers to use
+// AAD token credentials for the current user, but if that fails it will fall back to admin user credentials.
+func (crs *containerRegistryService) Credentials(
+	ctx context.Context, subscriptionId string, loginServer string,
+) (*DockerCredentials, error) {
```

Comment:
Out of curiosity: do you prefer pointer semantics here? I'm trying to learn about preferences and line of thinking.

Reply:
Thanks for engaging in the conversation.

I agree with the points about:
- Using `nil` to express optional (with documented behavior)
- Shared copy for mutation

To me, it seems that your concerns with token credentials refresh would be more adequately addressed generally by passing a credential object similar to `azcore.TokenCredential` with cache + fetch-on-demand semantics. A shared singleton could also be useful for application (non-library) development for practically. In that world however, it should be made *simply* singleton/global without additional ceremony.

The unaddressed concern for me is that if the team prefers pointer semantics, we should set clear guidelines and review context for nil-checking. Pointers are great, programmers are not as great as dealing with pointers, as our list of [nil panic related issues](https://github.com/Azure/azure-dev/issues?q=nil+%22panic%22+) grow.

Reply:
I really struggle here - TBH, I usually pick pointer semantics for two reasons:

1. We use pointer semantics in the codebase a lot.  And my sort of "boy scout" nature pushes me to follow the style of the rest of the codebase when possible.
2. In the case where a developer incorrectly tries to use the DockerCredential half of the return value when err is non nil, their program is more likely to crash via a panic due to a nil de-ref instead of just pulling the a field from the zero value, going a long for a bit and then failing because the zero value was "logically" invalid.  Obviously, something like a `Result<T>` would be nicer here (since GetResult or whatever could panic when the result models an error), but go doesn't have anything like that, so this is sort of a way to fake it, I guess?)

I know that (2) in practice is not something I should worry about, because when err is non nil you should just not expect the return value to be anything meaningful, but it's sort of like a belt and suspenders thing for me.

I really wish I had better heuristics here - I feel like sometimes in the go community the use of pointer vs value is also tied to the size of the underlying struct, but here a three word struct likely doesn't make much of a difference.